rfd/0-job-worker.md

```diff
@@ -0,0 +1,236 @@
+
+##### Job Execution
+
+The server will initially reexecute itself, using `exec.Command()`, as follows:
+
+```sh
+/proc/self/exe child [serve flags] -- command [args]...
+```
+
+When calling this, it will set the command's `SysProcAttr` with:
+
+```go
+syscall.SysProcAttr{
+    Cloneflags: syscall.CLONE_NEWPID | // New pid namespace
+        syscall.CLONE_NEWNS | // New mount namespace group
+        syscall.CLONE_NEWNET, // New network namespace
+    Unshareflags: syscall.CLONE_NEWNS, // Isolate process mounts from host
+}
+```
+
+In order to provide proper accounting of the job the server will internally track of the execution with:
+
+- buffers for `stdout` and `stderr`
+- process status (e.g. running, complete, error)
+- exit code
+
+Once reexecuted, the `child` command has several jobs to do. It has to remount the `/proc` filesystem, create and configure the cgroup, set the cgroup for the pid and fully replace the execution, using `syscall.Exec()`, with the job binary.
+
+##### Job Status
+
+Job status is extremely simple and only returns a status code and the exit_code. The status code is one of:
+
+- running: the job has not yet completed
+- completed: the job completed successfully
+- error: the job completed with a non-zero exit code
+- stopped: the job was manually stopped
+
+##### Job Output Storage
+
+Output for each job will be maintained, in memory, as long as the worker implementation exists. This poses a potential memory leak, but for the purposes of this challenge, is an acceptable solution. Later designs would likely need to write job output to disk or possibly even remote storage.
+
+##### Job Output Streaming
+
+The worker library streams job output through an `io.ReadCloser`. The server will indicate the end of the stream, if the job has completed, with the `io.EOF` error. The client may, at any time, close the stream to indicate that it is disconnecting. The client must always close the stream when it no longer needs it to prevent memory leaks.
+
+Internally, the library will maintain a buffer containing the job output. The job itself will be configured to use this buffer for `stdout` and `stderr` and the `Write()` method will be goroutine safe. When a client calls `JobOutput()` the library will create a new `io.Reader` that independently, and with goroutine safety, reads through to the end of the output buffer. If the job has already completed at this time, `io.EOF` will be returned. If not, subsequent calls to `Read()` will block until there is either more output to return or the job ends in which case `io.EOF` is returned.
```

Comment:
the buffer writer will signal the connected reader instances that there is more data. clients of the readers will have been blocking on the `Read()` call until that point.

i haven't totally landed on the implementation internally here, but there will definitely be goroutines and channels. i'm using the `io.Reader` interface externally, for the library, because it's idiomatic. i can certainly return a `←chan []byte` if that makes more sense. though, i don't think this is the reason for your question.

have i answered sufficiently?